In [45]:
%display latex

# MATH405:  Le problème d'Andrey Markov

## Exercice 1:

Pour prouver que $A = (a_{i,j})$ matrice stochastique implique que $1 \in Sp(A)$, on calcule

$$
^tA\begin{pmatrix}
1 \\
. \\
. \\
. \\
1
\end{pmatrix}
=
\begin{pmatrix}
\sum_{i=1}^n a_{i,1} \\
. \\
. \\
. \\
\sum_{i=1}^n a_{i,n}
\end{pmatrix}
=
\begin{pmatrix}
1 \\
. \\
. \\
. \\
1
\end{pmatrix}
$$

D'où la conclusion que 1 est dans le spectre de $^tA$ or on sait que le spectre de $A$ est égal au spectre de $^tA$.

## Exercice 2:

Soit $A = (a)_{i,j} \in M_n(\mathbb R), d = \begin{pmatrix} d_1 \\ \vdots \\ d_n \end{pmatrix} \in \mathbb R^n$. On a alors

$$
Ad = 
\begin{pmatrix}
a_{1,1} \cdots a_{1,n} \\
\vdots \ddots \vdots\\
a_{n, 1} \cdots a_{n,n}
\end{pmatrix} \begin{pmatrix} d1 \\ \vdots \\ d_n \end{pmatrix}
=
\begin{pmatrix}
d_1a_{1,1} + \cdots + d_na_{1,n} \\
\vdots \; \ddots \; \vdots \\
d_1a_{n,1} + \cdots + d_na_{n,n}
\end{pmatrix}
$$
On retrouve
$$
\sum_{j=1}^n d_1a_{j,1} + \cdots + d_na_{j,n}
=
d_1 \left( \sum_{i=1}^n a_{i,1} \right) + \cdots + d_n\left( \sum_{i=1}^n a_{i,n} \right)
= d_1 + \cdots + d_n = 1
$$

Donc on a $Ad$ qui est stochastique si $A$ et $d$ le sont aussi. 
Par hérédité immédiate on a $A^qd$ stochiastique $\forall q \in \mathbb N$ et donc une chaîne de Markov est une suite stochiastique.

## Exercice 3:

Pour prouver que toute matrice stochastique régulière remplit le critère de Perron-Frobenius, on cherche par contraposé:
Soit $M \in M_n(\mathbb R)$ stochastique telle que $\exists i \in \mathbb N, 1 \leq i \leq n$ tel que $ (M)_{i, \bullet} = 0$, on a donc $\forall q \in \mathbb N, \forall j \in \mathbb N, 1 \leq j \leq n$,
$$
(M^q)_{i,j} = (MM^{q-1})_{i,j} = (M)_{i, \bullet} (M^{q-1})_{\bullet, j}) = 0
$$

## Exercice 4:


La matrice stochastique $M$ représentant les changements de pièce de la souris où $M_{i,j}$ est la probabilité de passer de la pièce $j$ à la pièce $ i $ se décrit comme la matrice-ci dessous:

In [46]:
M = Matrix(QQ, [
    [1/3, 1/3, 0, 1/3, 0, 0, 0, 0, 0],
    [1/3, 1/3, 1/3, 0, 0, 0, 0, 0, 0],
    [0, 1/3, 1/3, 0, 1/3, 0, 0, 0, 0],
    [1/3, 0, 0, 1/3, 0, 0, 1/3, 0, 0],
    [0, 0, 1/4, 0, 1/4, 1/4, 0, 0, 1/4],
    [0, 0, 0, 0, 1/2, 1/2, 0, 0, 0],
    [0, 0, 0, 1/3, 0, 0, 1/3, 1/3, 0],
    [0, 0, 0, 0, 0, 0, 1/3, 1/3, 1/3],
    [0, 0, 0, 0, 1/3, 0, 0, 1/3, 1/3]
]).transpose()
M

[1/3 1/3   0 1/3   0   0   0   0   0]
[1/3 1/3 1/3   0   0   0   0   0   0]
[  0 1/3 1/3   0 1/4   0   0   0   0]
[1/3   0   0 1/3   0   0 1/3   0   0]
[  0   0 1/3   0 1/4 1/2   0   0 1/3]
[  0   0   0   0 1/4 1/2   0   0   0]
[  0   0   0 1/3   0   0 1/3 1/3   0]
[  0   0   0   0   0   0 1/3 1/3 1/3]
[  0   0   0   0 1/4   0   0 1/3 1/3]

On remarque que cette matrice peut être lue de deux façons:
- Si on la lit par colonne: les coéficients de la $j$-ème colomne nous indique la probabilité de passer à la $i$-ème pièce après la nuit.
- Si on la lit par ligne: les coéficients de la $i$-ème ligne indique la probabilité de provenir de la $j$-ème pièce.

### Partie 2:

On va démontrer que si on a $A_0$ le vecteur stochastique qui encode la probabilité pour la souris de se trouver dans une pièce la première nuit, alors $A_n = M^nA_0$.
On résonne par récurence.
On pose $A_0 = \left(\begin{smallmatrix} a_1^{(0)} \\ \vdots \\ a_9^{(0)} \end{smallmatrix}\right)$.
On a 
$$
MA_0 = \begin{pmatrix} 
P("1 \rightarrow 1"), \cdots, P("9 \rightarrow 1") \\
\vdots \ddots \vdots \\
P("1 \rightarrow 9"), \cdots, P("9 \rightarrow 9")
\end{pmatrix}
\begin{pmatrix}
P_0(1) \\
\vdots \ddots \vdots\\
P_0(n)
\end{pmatrix}
= \begin{pmatrix}
P_0(1)P("1 \rightarrow 1") + \cdots + P_0(n)P("n \rightarrow 1") \\
\vdots \ddots \vdots \\
P_0(1)P("1 \rightarrow n") + \cdots + P_0(n)P("n \rightarrow n")
\end{pmatrix}
= \begin{pmatrix}
P_1(1) \\
\vdots \\
P_1(n)
\end{pmatrix}
= A_1
$$
Soit $n \in \mathbb N$.
On suppose que $\forall n_0 \leq n, \; A_{n_0} = M^{n_0}A_0$.
On a
$$
MA_n = \begin{pmatrix}
P("1 \rightarrow 1"), \cdots, P("9 \rightarrow 1") \\
\vdots \ddots \vdots \\
P("1 \rightarrow 9"), \cdots, P("9 \rightarrow 9")
\end{pmatrix}
\begin{pmatrix}
P_n(1) \\
\vdots \\
P_n(n)
\end{pmatrix}
= \begin{pmatrix}
P_n(1)P("1 \rightarrow 1") + \cdots + P_n(n)P("n \rightarrow 1") \\
\vdots \ddots \vdots \\
P_n(1)P("1 \rightarrow n") + \cdots + P_n(n)P("n \rightarrow n")
\end{pmatrix}
= \begin{pmatrix}
P_{n+1}(1) \\
\vdots \\
P_{n+1}(n)
\end{pmatrix}
= A_{n+1}
$$
On a donc par récurence $A_n = M^nA_0$

On commence par supposer qu'il est équiprobable que la souris se trouve dans chaque pièce au départ, on encode cette information dans le vecteur ligne $I = \dfrac{1}{9}\left(\begin{smallmatrix} 1 \\ \vdots \\ 1 \end{smallmatrix}\right)$. On sait que sur une ligne $i$ on lit la probabilité que la souris passe de chaque pièce à la $i$-ème pièce. On peut récupérer cette information en multipliant M à gauche par un vecteur colonne $P_i$ nul sauf en la i-ème composante = 1. Par exemple pour i = 1, le vecteur est égal à $P_1 = \left(\begin{smallmatrix} 1, 0, \cdots, 0 \end{smallmatrix} \right)$. On a donc l'identitée:
$$
P(\text{"être dans la pièce }i\text{ au }n\text{-ième jour"}) = P_iM^nI
$$
Par exemple la probabilité de se trouver dans la pièce 1 le 15-ème jour est

In [57]:
n(Matrix(QQ, [[1,0,0,0,0,0,0,0,0]])*M**15*1/9*Matrix(QQ, [[1,1,1,1,1,1,1,1,1]]).transpose())

[0.110908804706547]

On a donc la probabilité que la souris se trouve dans la 1-ère pièce le 15-ème jour égale à 0.111 environ. 

### Partie 3:

Par définition de la régularité, pour prouver que $M$ est régulière, il suffit que toutes les lignes de M soit non nulles et de trouver un $\alpha \in \mathbb N$ tel que 
$$\forall (i,j) \in \mathbb N, 0 < i, j \leq n \implies \\ (M^\alpha)_{i, j} \neq 0$$

On remarque que trivialement la matrice n'a pas de colone nulles.
On remarque de la même façon que pour $\alpha = 15$ on a tous les $(M_{i, j}) \neq 0$.

On conclut que $M$ est régulière

### Partie 4:

On cherche à trouver le vecteur d'état stationnaire de $M$.
Pour cela on pourrait essayer de voir si en faisant tendre vers l'infini un vecteur choisis arbitrairement qu'on soumet à une chaine de Markov on ne verrait pas apparaitre une valeur que l'on connait. Cela cependant pourrait demander beaucoup de ressources sans resultat exacte.
On va donc plutôt utiliser la méthode par les vecteurs propres de $M$ pour trouver la valeur exacte. On sait que les vecteur propres $\mathbb R d$ associés à 1 sont solutions de $Md = d$, il nous suffit de trouver un vecteur propre représentant et un $\lambda$ tel que ce vecteur propre soit stochastique.

In [11]:
b = None
vecteurs_propres = M.eigenspaces_right()
for valeurs_propres, espaces_propres in vecteurs_propres:
    if valeurs_propres == 1:
        b = espaces_propres.basis()[0]
b

(1, 1, 1, 1, 4/3, 2/3, 1, 1, 1)

On divise le vecteur trouvé par la somme de ses coéficients pour avoir notre vecteur stationaire.

In [12]:
sum_b = 0
for k in range(9):
    sum_b += b[k]
b = b/sum_b
b

(1/9, 1/9, 1/9, 1/9, 4/27, 2/27, 1/9, 1/9, 1/9)

On conclut que le vecteur stationaire est le vecteur
$
\begin{pmatrix}
\frac{1}{9}, \frac{1}{9}, \frac{1}{9}, \frac{1}{9}, \frac{4}{27}, \frac{2}{27}, \frac{1}{9}, \frac{1}{9}, \frac{1}{9}
\end{pmatrix}
$

Ce vecteur correspond au choix de la loi qui régie la position initiale de la souris. Avec cette loi en particulier, peu importe le jour, on a une probabilité de tomber sur la souris dans une pièce chaques jours qui est fixe.